lark/grammars/lark.lark

```diff
@@ -7,46 +7,66 @@ _item: rule
      | token
      | statement
 
-rule: RULE rule_params priority? ":" expansions
-token: TOKEN token_params priority? ":" expansions
+rule: RULE_MODIFIERS? RULE rule_params priority? ":" rule_expansions
+token: TOKEN priority? ":" token_expansions
 
 rule_params: ["{" RULE ("," RULE)* "}"]
-token_params: ["{" TOKEN ("," TOKEN)* "}"]
 
 priority: "." NUMBER
 
-statement: "%ignore" expansions                    -> ignore
+statement: "%ignore" ignore_token                  -> ignore
          | "%import" import_path ["->" name]       -> import
          | "%import" import_path name_list         -> multi_import
          | "%override" rule                        -> override_rule
+         | "%override" token                       -> override_token
          | "%declare" name+                        -> declare
+         | "%extend" rule                          -> extend_rule
+         | "%extend" token                         -> extend_token
+
+ignore_token: ignore_item [ OP | "~" NUMBER [".." NUMBER]]
+ignore_item: STRING | TOKEN | REGEXP
 
 !import_path: "."? name ("." name)*
 name_list: "(" name ("," name)* ")"
 
-?expansions: alias (_VBAR alias)*
+?rule_expansions: rule_alias (_VBAR rule_alias)*
+
+?rule_inner_expansions: rule_expansion (_VBAR rule_expansion)*
+
+?rule_alias: rule_expansion ["->" RULE]
+
+?rule_expansion: rule_expr*
+
+?rule_expr: rule_atom [OP | "~" NUMBER [".." NUMBER]]
+?rule_atom: "(" rule_inner_expansions ")"
+          | "[" rule_inner_expansions "]" -> rule_maybe
+          | rule_value
+
+?rule_value: RULE "{" rule_value ("," rule_value)* "}" -> rule_template_usage
+           | RULE
+           | token_value
 
-?alias: expansion ["->" RULE]
+?token_expansions: token_expansion (_VBAR token_expansion)*
 
-?expansion: expr*
+?token_expansion: token_expr*
 
-?expr: atom [OP | "~" NUMBER [".." NUMBER]]
+?token_expr: token_atom [OP | "~" NUMBER [".." NUMBER]]
 
-?atom: "(" expansions ")"
-     | "[" expansions "]" -> maybe
-     | value
+?token_atom: "(" token_expansions ")"
+           | "[" token_expansions "]" -> token_maybe
+           | token_value
 
-?value: STRING ".." STRING -> literal_range
-      | name
-      | (REGEXP | STRING) -> literal
-      | name "{" value ("," value)* "}" -> template_usage
+?token_value: STRING ".." STRING -> literal_range
+            | TOKEN
+            | (REGEXP | STRING) -> literal
 
 name: RULE
     | TOKEN
 
 _VBAR: _NL? "|"
 OP: /[+*]|[?](?![a-z])/
-RULE: /!?[_?]?[a-z][_a-z0-9]*/
+RULE: /_?[a-z][_a-z0-9]*/
+RULE_MODIFIERS: /!|![?](?=[a-z])|[?]!?(?=[a-z])/
```

Comment:
Why do you need the `?=`s here? I don't recall that ? or ! are valid anywhere else in the grammar..?

Reply:
Oh I see. But wouldn't users be able to bypass it by adding whitespace between them?

Reply:
Drat!  I missed a test case.  But no, this test fails similarly in `load_grammar.py`, the original `lark.lark`, and my modified version:

```        g = """start: rule1
            rule1: rule2
            ?! rule2: "a"
        """
```

`load_grammar.py`:
```
Traceback (most recent call last):
  File "C:\Ross\Source\lark\lark\parsers\lalr_parser_state.py", line 77, in feed_token
    action, arg = states[state][token.type]
KeyError: 'OP'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Ross\Source\lark\tests\test_lark_lark.py", line 169, in test_11_rule_modifier_query_bang_space_lg
    Lark(g)
...
  File "C:\Ross\Source\lark\lark\lexer.py", line 674, in lex
    raise UnexpectedToken(token, e.allowed, state=parser_state, token_history=[last_token], terminals_by_name=self.root_lexer.terminals_by_name)
lark.exceptions.UnexpectedToken: Unexpected token Token('OP', '?') at line 3, column 13.
Expected one of:
        * "%declare"
        * "%import"
        * "%extend"
        * RULE
        * "%override"
        * _NL
        * $END
        * "%ignore"
        * TOKEN
        * RULE_MODIFIERS
Previous tokens: [Token('_NL', '\n            ')]
```

Original `lark.lark`:
```
Traceback (most recent call last):
  File "C:\Ross\Source\lark\lark\lexer.py", line 665, in lex
    yield lexer.next_token(lexer_state, parser_state)
  File "C:\Ross\Source\lark\lark\lexer.py", line 598, in next_token
    raise UnexpectedCharacters(lex_state.text, line_ctr.char_pos, line_ctr.line, line_ctr.column,
lark.exceptions.UnexpectedCharacters: No terminal matches '?' in the current parser context, at line 3 col 13

            ?! rule2: "a"
            ^
...
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Ross\Source\lark\tests\test_lark_lark.py", line 176, in test_11_rule_modifier_query_bang_space_ll
    self.lark_parser.parse(g)
...
  File "C:\Ross\Source\lark\lark\lexer.py", line 674, in lex
    raise UnexpectedToken(token, e.allowed, state=parser_state, token_history=[last_token], terminals_by_name=self.root_lexer.terminals_by_name)
lark.exceptions.UnexpectedToken: Unexpected token Token('OP', '?') at line 3, column 13.
Expected one of:
        * _NL
        * "%import"
        * "%ignore"
        * "%override"
        * TOKEN
        * $END
        * "%declare"
        * RULE
Previous tokens: [Token('_NL', '\n            ')]
```

Modified `lark.lark`:
```
Traceback (most recent call last):
  File "C:\Ross\Source\lark\lark\lexer.py", line 665, in lex
    yield lexer.next_token(lexer_state, parser_state)
  File "C:\Ross\Source\lark\lark\lexer.py", line 598, in next_token
    raise UnexpectedCharacters(lex_state.text, line_ctr.char_pos, line_ctr.line, line_ctr.column,
lark.exceptions.UnexpectedCharacters: No terminal matches '?' in the current parser context, at line 3 col 13

            ?! rule2: "a"
            ^
...
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Ross\Source\lark\tests\test_lark_lark.py", line 176, in test_11_rule_modifier_query_bang_space_ll
    self.lark_parser.parse(g)
...
  File "C:\Ross\Source\lark\lark\lexer.py", line 674, in lex
    raise UnexpectedToken(token, e.allowed, state=parser_state, token_history=[last_token], terminals_by_name=self.root_lexer.terminals_by_name)
lark.exceptions.UnexpectedToken: Unexpected token Token('OP', '?') at line 3, column 13.
Expected one of:
        * "%declare"
        * RULE
        * RULE_MODIFIERS
        * $END
        * "%ignore"
        * "%extend"
        * "%import"
        * TOKEN
        * _NL
        * "%override"
Previous tokens: [Token('_NL', '\n            ')]
```